File: packages/js/http/src/prepared-order-store.ts
/**
 * Prepared-order amount-authority persistence for the shipped HTTP handler.
 *
 * POST /prepare persists `{ amount, summary?, metadata? }` under `host_order:<orderId>`.
 * POST /checkouts and GET /orders/{id}/summary read that row back.
 */

export const PREPARED_ORDER_META_PREFIX = "host_order:" as const;

export interface PreparedOrderMetaRow {
  readonly value: string;
  readonly rev: number;
}

export interface PreparedOrderMetaStore {
  getMeta(key: string): PreparedOrderMetaRow | undefined | Promise<PreparedOrderMetaRow | undefined>;
  casMeta(
    key: string,
    value: string,
    expectedRev: number | null,
  ):
    | { status: "ok" | "conflict"; row: PreparedOrderMetaRow }
    | Promise<{ status: "ok" | "conflict"; row: PreparedOrderMetaRow }>;
}

export type PreparedOrderAmount =
  | { readonly sats: number | string; readonly currency?: never; readonly value?: never }
  | {
      readonly currency: string;
      readonly value: string;
      readonly sats?: never;
    };

export interface StoredPreparedOrder {
  readonly amount: PreparedOrderAmount;
  readonly summary?: unknown;
  readonly metadata?: Record<string, unknown>;
}

export interface PreparedOrderStore {
  persist(orderId: string, stored: StoredPreparedOrder): Promise<void>;
  read(orderId: string): Promise<StoredPreparedOrder | null>;
}

export function createPreparedOrderStore(
  store: PreparedOrderMetaStore,
  prefix: string = PREPARED_ORDER_META_PREFIX,
): PreparedOrderStore {
  if (prefix.length === 0) {
    throw new Error("createPreparedOrderStore: prefix must be non-empty");
  }

  function metaKey(orderId: string): string {
    if (typeof orderId !== "string" || orderId.length === 0) {
      throw new TypeError("createPreparedOrderStore: orderId must be a non-empty string");
    }
    return `${prefix}${orderId}`;
  }

  return {
    async persist(orderId, stored) {
      const key = metaKey(orderId);
      const payload = JSON.stringify(stored);
      const current = await store.getMeta(key);
      const first = await store.casMeta(key, payload, current === undefined ? null : current.rev);
      if (first.status === "ok") return;
      const again = await store.getMeta(key);
      const second = await store.casMeta(key, payload, again === undefined ? null : again.rev);
      if (second.status !== "ok") {
        throw new Error("createPreparedOrderStore: failed to persist prepared order");
      }
    },
    async read(orderId) {
      const row = await store.getMeta(metaKey(orderId));
      if (row === undefined) return null;
      try {
        const parsed = JSON.parse(row.value) as unknown;
        if (!isStoredPreparedOrder(parsed)) return null;
        return parsed;
      } catch {
        return null;
      }
    },
  };
}

function isStoredPreparedOrder(value: unknown): value is StoredPreparedOrder {
  if (typeof value !== "object" || value === null) return false;
  const record = value as Record<string, unknown>;
  if (typeof record.amount !== "object" || record.amount === null) return false;
  const amount = record.amount as Record<string, unknown>;
  if (typeof amount.sats === "string" || typeof amount.sats === "number") return true;
  return typeof amount.currency === "string" && typeof amount.value === "string";
}

File: packages/ruby/openreceive-server/lib/openreceive/server/prepared_order_store.rb
# frozen_string_literal: true

require "json"

module OpenReceive
  module Server
    # Prepared-order amount-authority persistence — Ruby port of
    # packages/js/http/src/prepared-order-store.ts.
    #
    # POST /prepare persists `{ amount, summary?, metadata? }` under `host_order:<orderId>`.
    # POST /checkouts and GET /orders/{id}/summary read that row back.
    class PreparedOrderStore
      PREPARED_ORDER_META_PREFIX = "host_order:"

      def initialize(store, prefix: PREPARED_ORDER_META_PREFIX)
        raise ArgumentError, "PreparedOrderStore: prefix must be non-empty" if prefix.to_s.empty?

        @store = store
        @prefix = prefix.to_s
      end

      def meta_key(order_id)
        raise ArgumentError, "PreparedOrderStore: order_id must be a non-empty string" unless order_id.is_a?(String) && !order_id.empty?

        "#{@prefix}#{order_id}"
      end

      # Upsert: read current rev, cas, retry once on conflict (mirrors JS persist).
      def persist(order_id, stored)
        key = meta_key(order_id)
        payload = JSON.generate(stored)
        current = @store.get_meta(key)
        first = @store.cas_meta(key: key, value: payload, expected_rev: current.nil? ? nil : current["rev"])
        return if first["status"] == "ok"

        again = @store.get_meta(key)
        second = @store.cas_meta(key: key, value: payload, expected_rev: again.nil? ? nil : again["rev"])
        return if second["status"] == "ok"

        raise "PreparedOrderStore: failed to persist prepared order"
      end

      # Returns the stored hash or nil when missing / malformed.
      def read(order_id)
        row = @store.get_meta(meta_key(order_id))
        return nil if row.nil?

        begin
          parsed = JSON.parse(row["value"])
        rescue JSON::ParserError
          return nil
        end
        return nil unless stored_prepared_order?(parsed)

        parsed
      end

      private

      def stored_prepared_order?(value)
        return false unless value.is_a?(Hash)

        amount = value["amount"]
        return false unless amount.is_a?(Hash)

        sats = amount["sats"]
        return true if sats.is_a?(String) || sats.is_a?(Numeric)

        amount["currency"].is_a?(String) && amount["value"].is_a?(String)
      end
    end
  end
end
